plugin/trino-memory/src/main/java/io/trino/plugin/memory/MemoryMetadata.java

```diff
@@ -115,9 +116,29 @@ public synchronized void dropSchema(ConnectorSession session, String schemaName)
             throw new TrinoException(NOT_FOUND, format("Schema [%s] does not exist", schemaName));
         }
 
+        // DropSchemaTask has the same logic, but needs to check in connector side considering concurrent operations
+        if (!isSchemaEmpty(schemaName)) {
+            throw new TrinoException(SCHEMA_NOT_EMPTY, "Schema not empty: " + schemaName);
+        }
+
         verify(schemas.remove(schemaName));
     }
 
+    private boolean isSchemaEmpty(String schemaName)
+    {
+        if (tables.values().stream()
```

Comment:
Is this thread-safe? It's backed by a normal HashMap and there is no synchronisation.

Reply:
It should be. https://docs.oracle.com/javase/tutorial/essential/concurrency/syncmeth.html

> First, it is not possible for two invocations of synchronized methods on the same object to interleave. When one thread is executing a synchronized method for an object, all other threads that invoke synchronized methods for the same object block (suspend execution) until the first thread is done with the object.